framework/src/modules/interoperability/utils.ts

```diff
@@ -106,3 +146,295 @@ export const computeValidatorsHash = (
+			};
+		}
+
+		// Last certificate height should be less than new certificate height
+		if (decodedCertificate.height <= partnerChainAccount.lastCertificate.height) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: new Error('Certificate height should be greater than last certificate height.'),
+			};
+		}
+	}
+
+	return {
+		status: VerifyStatus.OK,
+	};
+};
+
+export const checkActiveValidatorsUpdate = (
+	txParams: CrossChainUpdateTransactionParams,
+): VerificationResult => {
+	if (
+		txParams.activeValidatorsUpdate.length !== 0 ||
+		txParams.newCertificateThreshold > BigInt(0)
+	) {
+		// Non-empty certificate
+		if (txParams.certificate.equals(EMPTY_BYTES)) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: new Error(
+					'Certificate cannot be empty when activeValidatorsUpdate is non-empty or newCertificateThreshold >0.',
+				),
+			};
+		}
+
+		// params.activeValidatorsUpdate has the correct format
+		for (let i = 0; i < txParams.activeValidatorsUpdate.length; i += 1) {
+			const currentValidator = txParams.activeValidatorsUpdate[i];
+			const nextValidator = txParams.activeValidatorsUpdate[i + 1];
+			if (currentValidator.blsKey.byteLength !== VALID_BLS_KEY_LENGTH) {
+				return {
+					status: VerifyStatus.FAIL,
+					error: new Error(`BlsKey length should be equal to ${VALID_BLS_KEY_LENGTH}.`),
+				};
+			}
+			if (currentValidator.blsKey.compare(nextValidator.blsKey) > -1) {
+				return {
+					status: VerifyStatus.FAIL,
+					error: new Error('Validators blsKeys must be unique and lexicographically ordered'),
+				};
+			}
+		}
+	}
+
+	return {
+		status: VerifyStatus.OK,
+	};
+};
+
+export const checkInboxUpdateValidity = (
+	txParams: CrossChainUpdateTransactionParams,
+	partnerChannelData: ChannelData,
+): VerificationResult => {
+	const partnerChainIDBuffer = getIDAsKeyForStore(txParams.sendingChainID);
+	const decodedCertificate = codec.decode<Certificate>(certificateSchema, txParams.certificate);
+	const { crossChainMessages, messageWitness, outboxRootWitness } = txParams.inboxUpdate;
+	const ccmHashes = crossChainMessages.map(ccm => hash(ccm));
+
+	let newInboxRoot;
+	let newInboxAppendPath = partnerChannelData.inbox.appendPath;
+	let newInboxSize = partnerChannelData.inbox.size;
+	for (const ccm of ccmHashes) {
+		const { appendPath, size } = regularMerkleTree.calculateMerkleRoot({
+			value: ccm,
+			appendPath: newInboxAppendPath,
+			size: newInboxSize,
+		});
+		newInboxAppendPath = appendPath;
+		newInboxSize = size;
+	}
+	// If inboxUpdate contains a non-empty messageWitness, then update newInboxRoot to the output
+	if (
+		!txParams.certificate.equals(EMPTY_BYTES) &&
+		txParams.inboxUpdate.messageWitness.siblingHashes.length !== 0 &&
+		txParams.inboxUpdate.messageWitness.partnerChainOutboxSize > 0
+	) {
+		newInboxRoot = regularMerkleTree.calculateRootFromRightWitness(
+			newInboxSize,
+			newInboxAppendPath,
+			messageWitness.siblingHashes,
+		);
+
+		const proof = {
```

Comment:
below here, I think it should be outside of the `messageWitness` if clause

Reply:
```
If params contains a non-empty certificate and an inboxUpdate, the validity of the latter is checked by:
```
where if inboxUpdate is non-empty then it should also have non-empty messageWitness as its a required property on schema. So I did a check for this whole block on top of this func `isInboxUpdateEmpty` where it checks for all the required properties of inboxUpdate.

so I assume, 
```
Then validate the newly updated root against the certificate state root using the provided outboxRootWitness. Using notation from [LIP 0039](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0039.md), this is done via the function verify(queryKeys, proof, certificate.stateRoot) with:
```
is also implying that inboxUpdate as well as messageWitness should exist.

Reply:
I think it's possible to have non-empty inbox update and empty message witness if I remember correctly.
I think similar check is done [here](https://github.com/LiskHQ/lisk-sdk/blob/9b722082fee3da23db5b514cc4da12555b6cf005/framework/src/modules/interoperability/utils.ts#L454-L457)